server/remote_cache/cachetools/cachetools.go

```diff
@@ -269,6 +276,232 @@ func UploadProtoToCAS(ctx context.Context, cache interfaces.Cache, instanceName
 	return uploadProtoToCache(ctx, cas, instanceName, in)
 }
 
+// BatchCASUploader uploads many files to CAS concurrently, batching small
+// uploads together and falling back to bytestream uploads for large files.
+type BatchCASUploader struct {
+	ctx              context.Context
+	eg               *errgroup.Group
+	cache            interfaces.Cache
+	bsClient         bspb.ByteStreamClient
+	casClient        repb.ContentAddressableStorageClient
+	req              *repb.BatchUpdateBlobsRequest
+	currentBatchSize int64
+	instanceName     string
+	uploads          map[digest.Key]struct{}
+}
+
+func NewBatchCASUploader(ctx context.Context, env environment.Env, instanceName string) (*BatchCASUploader, error) {
```

Comment:
nit: maybe add a comment that this should only be kept around for the life of a request (since you're capturing context?)